internal/graph/graph.go

```diff
@@ -385,3 +385,43 @@ func (g *RelationshipGraph) getRelationshipEdgesWithTargetRewrite(
 		panic("unexpected userset rewrite encountered")
 	}
 }
+
+// NewLayeredCheckResolver constructs a CheckResolver that is composed of various CheckResolver layers.
+// Specifically, it constructs a CheckResolver with the following composition:
+//
+//	CycleDetectionCheckResolver  <-----|
+//		CachedCheckResolver              |
+//			LocalChecker                   |
+//				CycleDetectionCheckResolver -|
+//
+// The returned CheckResolverCloser should be used to close all resolvers involved in the
+// composition after you are done with the CheckResolver.
+func NewLayeredCheckResolver(
+	localResolverOpts []LocalCheckerOption,
+	cacheEnabled bool,
+	cachedResolverOpts []CachedCheckResolverOpt,
+) (CheckResolver, CheckResolverCloser) {
```

Comment:
This looks like a code smell. `CheckResolver` interface includes a `Close` method, so why do we need `CheckResolverCloser`??

Reply:
The intent was that anything that implements the `Close` method of the CheckResolver implements the `CheckResolverCloser` signature. It's a way to share/communicate that shared behavior.

We can refactor to simply `func()` though. I just prefer more strongly typed signatures, and giving that signature a typed name makes it as such.

Reply:
i don't follow. Why do we need the `CheckResolverCloser` interface in the first place?